### DISCUSSION TITLE
2d platformer game

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#1930] [Show and tell]
Hello,
here is the link to the video of my last project.  I was going to post it in cocos2dx forum but I think there is no point there.
Anyway, check it out in the link bellow.
Cheers

https://www.youtube.com/watch?v=3nPVrxiO-bc

**Comment by paulocoutinhox** [comment]
Wow man. Can you share the source? It has a lot of mechanics to everyone build platform games. Im opening a lot of code of Axmol projects in my repo to make easy to people adopt Axmol and understand how to build things with it.

**Reply by johnspeny** [reply]
Gives me blasphemous vibes, Looking nice I could give you a hand if you need

**Reply by rudiHammad** [reply]
I am just looking for composers, anything else I got it covered, thanks.
And yup, a little chubby green penitent.

**Comment by GamesStudioTwo** [comment]
fantastic, I can see you got some mechanics in there. is all the artwork yours ? its great thanks for the show.

**Reply by rudiHammad** [reply]
thx. Yes, I do everything except music. I'll hire a compositor if I can afford it .

**Reply by paulocoutinhox** [reply]
You can use iA music, have some free services that you can generate and download.

**Reply by GamesStudioTwo** [reply]
That's fantastic to hear some very good artwork love the parallax scrolling, I'm great a blender myself and making 3d stuff but rubbish at 2d sprites. For music just look up royalty free music.
there's lost of free great stuff out there just give them a shout in ya credits. problem with music is  you start listing and after hearing about 15 tunes then you become tone deaf and noting matches what you want music is harder than programming. I noticed you don't have any ropes to swing on, are you using box2d. I will be happy to point you in the right direction. I'm just putting in ropes myself and using Tiled what a great program.

**Reply by rudiHammad** [reply]
The thing is that I don't really like A.I music because they feel soul less to me. And royality free music is an option, but you can't really do much about it. What you hear is what you get. I prefer to save some money and support composers. Plus you can talk with them and discuss what you are looking for.
I didn't use any physics engine. I wrote my own, which is basically a good old tile collision system.
Cheers

**Reply by johnspeny** [reply]
@rudiHammad how long have you been developing the game now. Its a genre am a fan of, i guess its metroidvania right and they take lots of time. The next challenge i believe will be the map that shows your current location etc like in blasphemous or last faith

**Reply by rudiHammad** [reply]
Around 10 months. What is most time consuming is art really.
Before getting into more metroidvania features I want focus on more core generic stuff. For instance, how to create slots and save your progress; how to poperly setup a gamepad; how to encrypt your assets to make it harder to hack; how to setup an online lead bord; etc..
When all this is clear I'll come back to the game.

**Reply by GamesStudioTwo** [reply]
I don't know if this is what your looking. for Saving parts of game you can use the UserDefault Class. its easy to use
auto hero = UserDefault::getInstance();
        hero->setStringForKey("Level","Map01.tmx");
        hero->setIntegerForKey("Score",0);
       hero->setIntegerForKey("Lives",3);
       hero->setBoolForKey("Magic Potions1",true);

    auto mapName = hero->getStringForKey("Level");

for gamepad is it hardware gamepad you talking about or is it for mobile app's
for encryption I only know sprite sheet's using Texture Packers but you using a tilemap.

**Reply by rudiHammad** [reply]
I didn't investigate that yet, so maybe? I don't know. My target is PC not mobile. What I mean is something like this 
![Capture](https://github.com/axmolengine/axmol/assets/25636976/313a82a6-f89c-47e4-b53e-a862cf80759b)
This is hollow knight for pc. There are four slots. This is great if you have brothers/sisters to have each one have it's own. It also shows you the % of progress of your game, hours played, etc.. 

Same goes for gamepad, i am referring to an xbox controller, or ps5, etc...
For encryption, yeah, I mean all assets, tiles, textures... event data file to prevent hacking for instance the scores file and give your self a million points. 

Each one of those things are big topic on their one, so maybe it's something to talk about in separate discussions for all the community.
I don't know how many developers are interested in developing for PC, but it my case that's my main target. I'll also do mobil, but first PC because I want to sell on steam.

**Reply by rh101** [reply]
Wouldn't it be better to have separate player profiles, each with their own list of saved game slots to use?  When the user opens the game, it would ask them which profile they want to open/log into etc., and then it loads into the game in that profile, so all saved games belong to that specific profile.  Personally, I feel that it would be a much better user experience.

For actually saving progress, that's really up to you.  You could save each slot data in a separate file, so if one is somehow corrupted, it won't affect the rest.  Alternatively, you could use SQLite or some other data format for that etc..

If you're saving in separate files, and have separate player profiles, then it makes it simple, since you can structure the folders like this:
`profiles/<profile_id>/<saved_games>/slot_0.data` etc.

There is only so much you can do to encrypt resources, because textures can always be pulled directly from the video memory, since they're not encrypted in there.  The encrypted format in Axmol would be sufficient, but it won't deter anyone determined and knowledgeable enough.  Same goes for the saved games and score files, you really can't do much about it.  The only way to prevent this kind of hacking is to store the results on a remote server, including checking all values coming from the client to update those values (never trust the client!).

For tile asset files, you could convert the file data into your own custom format, and then code up a reader that parses that data then passes it to the tiled implementation in Axmol.  It can be enough of a deterrent for most, since it makes it much more time consuming to figure out the format and then convert it back or adapt it in some way.  This actually applies to any kind of data in your asset folder, but it would mean you need to invest time into coding it all up.

**Reply by GamesStudioTwo** [reply]
That's what I have its a custom tile map reader that take a spritesheet using texturpacker also save on memory space
tile's a just 0000 to 1023 I use python script to cut out the tile's into 64by64 removing blanks naming them 0000 to 1023.

**Reply by rudiHammad** [reply]
> Wouldn't it be better to have separate player profiles, each with their own list of saved game slots to use? When the user opens the game, it would ask them which profile they want to open/log into etc., and then it loads into the game in that profile, so all saved games belong to that specific profile. Personally, I feel that it would be a much better user experience.
> 
Yes, you're right. Usually someone else would have it's own separate profile, and inside of it the different slots. 
Ok cool, I won't stress out too much about encryption then. Thanks for the tips.

**Reply by rudiHammad** [reply]
By the way I found this about saving/loading data which looks alright.
https://www.youtube.com/watch?v=IEiOP7Y-Mbc&t=1505s
